torch/__init__.py

```diff
@@ -300,10 +300,69 @@ def set_default_dtype(d):
 
 def set_deterministic(d):
-    r"""Sets a global flag to force all operations to use a deterministic
-    implementation if available. If an operation that does not have a
-    deterministic implementation is called while this setting is True, the
-    operation will throw a RuntimeError.
+    r"""Sets a global flag to force normally nondeterministic operations to
+    use a deterministic implementation, if available. Or if any of these
+    operations do not have a deterministic implementation, they will throw
+    a :class:`RuntimeError`.
+
+    .. warning::
+        This feature is a work in progress, so it does not affect every
+        nondeterministic operation yet. All affected operations are listed below.
+        Any other operations in PyTorch that call the listed operations
+        internally are also affected.
+
+    The following normally-nondeterministic operations will act
+    deterministically when `d=True`:
+
+        * Convolution operations
+        * :func:`torch.bmm` for sparse-dense CUDA inputs
+
+    The following normally-nondeterministic operations will throw a
+    :class:`RuntimeError` when `d=True`:
+
+        * Pooling backward functions for CUDA
+            - :class:`torch.nn.AvgPool3d`
+            - :class:`torch.nn.AdaptiveAvgPool2d`
+            - :class:`torch.nn.AdaptiveAvgPool3d`
+            - :class:`torch.nn.MaxPool3d`
+            - :class:`torch.nn.AdaptiveMaxPool2d`
+            - :class:`torch.nn.FractionalMaxPool2d`
+            - :class:`torch.nn.FractionalMaxPool3d`
+
+        * Upsampling backward functions of :func:`torch.nn.functional.interpolate` for CUDA with modes:
+            - `linear`
+            - `bilinear`
+            - `bicubic`
+            - `trilinear`
+
+        * Padding backward functions for CUDA
+            - :class:`torch.nn.ReflectionPad1d`
+            - :class:`torch.nn.ReflectionPad2d`
+            - :class:`torch.nn.ReplicationPad1d`
+            - :class:`torch.nn.ReplicationPad2d`
+            - :class:`torch.nn.ReplicationPad3d`
+
+        * Loss backward functions for CUDA
+            - :class:`torch.nn.NLLLoss`
+            - :class:`torch.nn.CTCLoss`
+
+        * :class:`torch.nn.EmbeddingBag` backward function for CUDA
+        * :func:`torch.scatter_add_` for CUDA
+        * :func:`torch.index_add_` for CUDA
+        * :func:`torch.index_select` backward function for CUDA
+        * :func:`torch.repeat_interleave` backward function for CUDA
+        * :func:`torch.histc` for CUDA
+        * :func:`torch.bincount` for CUDA
+
+    A handful of CUDA operations are nondeterministic if the CUDA version is
+    10.2 or greater, unless the environment variable `CUBLAS_WORKSPACE_CONFIG=:4096:8`
+    or `CUBLAS_WORKSPACE_CONFIG=:16:8` is set. If one of these environment variable
+    configurations is not set, a :class:`RuntimeError` will be raised from these
+    operations:
+
+        * :func:`torch.mm`
+        * :func:`torch.mv`
+        * :func:`torch.bmm`
```

Comment:
It would be good to help pave the strategy for updating this docstring when we add more deterministic checks. It doesn't have to be done this PR, but good start might be to add a comment to all of the determinism check sites along the lines of "hey, if you copy paste me, you need to also update something else, go read Note [Determinism note] for more details". Adding this to the API docs for the determinism functions is also a good idea.

Reply:
Good idea, I'll follow up in another PR.